site/content/blog/2021-04-23-optimizing-go-compiler-2.md

```diff
@@ -0,0 +1,343 @@
+
+Once we have our values initialized, we can study the interesting instructions, those in block `b43`:
+
+```
+v18 (+7) = LoweredNilCheck <void> v7 v1
+v29 (7) = InlMark <void> [0] v1
+v152 (+8) = InlMark <void> [1] v1
+v34 (+79) = MOVQload <uint64> v7 v1
+v171 (+84) = MOVBstore <mem> v248 v34 v1
+v168 (+85) = SHRQconst <uint64> [8] v34
+v216 (+89) = SHRQconst <uint64> [40] v34
+v180 (+91) = SHRQconst <uint64> [56] v34
+v219 (88) = MOVLstore <mem> [1] v248 v168 v171
+v243 (90) = MOVWstore <mem> [5] v248 v216 v219
+v255 (91) = MOVBstore <mem> [7] v248 v180 v243
+```
+
+We can skip the first three lines, which are again formalities, and start on the one for the value `v34`:
+
+```
+v34 (+79) = MOVQload <uint64> v7 v1
+```
+
+Cool, our first real instruction, `MOVQload`! Now, what does that cryptic name means? The best way to know is to go directly to [its definition](https://github.com/golang/go/blob/7240a18adbfcff5cfe750a1fa4af0fd42ade4381/src/cmd/compile/internal/ssa/gen/AMD64Ops.go#L702):
+
+```go
+// load 8 bytes from arg0+auxint+aux. arg1=mem
+{
+	name: "MOVQload",
+	argLength: 2,
+	reg: gpload,
+	asm: "MOVQ",
+	aux: "SymOff",
+	typ: "UInt64",
+	faultOnNilArg0: true,
+	symEffect: "Read",
+}
+```
+
+The data in this struct defines what the `MOVQload` instruction does; the corresponding code in assembly, `MOVQ`; its returning type, `UInt64`; or the number of arguments it receives, 2. The comment does a pretty good job explaining what the instruction does: it loads 8 bytes from the data pointed to by the first argument (plus some auxiliary arguments, if present, which we will see are constant numbers in square brackets) into the second, which should represent the memory. If we go back to our line,
+
+```
+v34 (+79) = MOVQload <uint64> v7 v1
+━━┓                           ━┓ ━┓
+  ┗ we can see this            ┃  ┗ memory
+    one as temp                ┗ src
+```
+
+we now can understand that this loads the contents from the `src` argument (remember that `v7`represented`src`), into the memory we initialized at the very beginning (represented by `v1`). There are no auxiliary arguments, so we can safely forget about those! The result of this instruction is represented by the `v34` value, which we can understand as the `temp` variable.
+
+This first line of the block, then, seems perfectly fine: it loads the whole contents of the `src` array into memory (which is effectively the `temp` variable we defined in the code), and it does it with a single instruction. We can no longer optimize this.
+
+But remember that the original function did two things: first, it loaded the contents of the `src` array into a temporal variable, and then it stored those contents into the `dest` slice. We have already loaded the contents into memory with the line we discussed above, so the rest of the block should do the rest of the work: store those contents from memory into the `dest` slice (which was represented by value `v248`). Let's see the rest of the block again:
```

Comment:
temporal

Reply:
Same thing as with destiny/destination. We reuse a lot of words in Spanish, I guess, and that shows :joy: Fixed in https://github.com/mattermost/mattermost-developer-documentation/pull/803/commits/bcd613fd20785b2de630c486a25b70caa5dfc2fe, thanks!